### PR TITLE
[WIP] Implement energy and power classes to allow appropriate units to be used everywhere, separate batteries from magazines

### PR DESCRIPTION
**Pull request by ymber**
#### Summary
```SUMMARY: Infrastructure "Implement energy and power classes, separate batteries from magazines"```

#### Purpose of change
The solution I've got in #30656 is kinda flaky because it has to work with integer kJ values for items with power draw <1W. Using `energy_quantity` and `power` will allow millijoule precision and they support "10 mW" style energy and power strings for json definitions.

#### Describe the solution
Implement an `energy_quantity` class similar to `time_duration`. Implement a `power` class for electrical power. Separate magazines and batteries for better SoC.

- [x] Implement `energy_quantity` class
- [x] Implement new item type BATTERY with `energy_quantity`
- [ ] Allow reloading items with BATTERY items
- [ ] Rework ammo consumption to work with BATTERY items
- [ ] Implement `power` class
- [ ] Use `power` for items with `power_draw`
- [ ] Use `power` for vehicle power

#### Describe alternatives you've considered
Making battery charges millijoules would be the easiest solution but then we'd have to use that precision everywhere and some cases are more appropriately handled with larger units.

### CODE REVIEW

**Comment by Inglonias** [comment]
Are you planning to adjust vehicle batteries in the same way as the personal ones?

**Comment by ymber** [comment]
That's the plan. Anything electric gets the new toys.

**Comment by jbytheway** [comment]
You might want to try to split this up into smaller PRs; this seems like a very large project to swallow in one piece.

File: src/energy.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+#define ENERGY_H
+
+#include <utility>
+
+#include "rng.h"
+
+class JsonIn;
+class JsonOut;
+
+class energy_quantity
```
Comment (BevapDin) [inline src/energy.h]: What's the advantage of this class over the existing `units::quantity` template class?

Comment (ymber) [reply src/energy.h]: None really, I just hadn't considered using `units::quantity`.

**Comment by ymber** [comment]
The scale of this change is far larger than originally expected and it would be way too big to review and test properly at once. The new tracking issue is #32067 and the implementation will be split up into smaller chunks.

**Pull request closed without merging**